pkg/scheduler/internal/cache/cache.go

```diff
@@ -513,15 +513,34 @@ func (cache *schedulerCache) UpdatePod(oldPod, newPod *v1.Pod) error {
 	// An assumed pod won't have Update/Remove event. It needs to have Add event
 	// before Update event, in which case the state would change from Assumed to Added.
 	case ok && !cache.assumedPods.Has(key):
-		if currState.pod.Spec.NodeName != newPod.Spec.NodeName {
-			klog.ErrorS(nil, "Pod updated on a different node than previously added to", "pod", klog.KObj(oldPod))
-			klog.ErrorS(nil, "SchedulerCache is corrupted and can badly affect scheduling decisions")
-			os.Exit(1)
-		}
-		if err := cache.updatePod(oldPod, newPod); err != nil {
-			return err
+		// A Pod delete event followed by an immediate Pod add event may be merged
+		// into a Pod update event. In this case, we should invalidate the old Pod, and
+		// then add the new Pod.
+		if oldPod.UID != newPod.UID {
+			newKey, err := framework.GetPodKey(newPod)
```

Comment:
It looks like we are duplicating code. I would like to avoid that.

I think the functions `removePod`, `addPod`, `updatePod` are badly named. They should be `removePodFromNode`, `addPodToNode` and `updatePodInNode` respectively. Then, we can have new functions for `addPod` and `removePod` that include handling the `podStates` and `assumedPods`.

Reply:
I agree with you.

At the same time, I *prefer* to keep the operations of `assumedPods` outside these `addPod` / `removePod` / `updatePod` functions. Emm, I think it would be a little strange to handle `assumedPod` in these functions.

And I have finished the code refactoring, PTAL ^ ^

Reply:
Not really, having them in there would help consistency.

Reply:
@alculquicondor done, PTAL.